File: ta4j-core/src/test/java/org/ta4j/core/indicators/adx/PlusDMIndicatorTest.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.indicators.adx;

import static org.ta4j.core.TestUtils.assertNumEquals;

import org.junit.Test;
import org.ta4j.core.Indicator;
import org.ta4j.core.indicators.AbstractIndicatorTest;
import org.ta4j.core.mocks.MockBarSeriesBuilder;
import org.ta4j.core.num.Num;
import org.ta4j.core.num.NumFactory;

public class PlusDMIndicatorTest extends AbstractIndicatorTest<Indicator<Num>, Num> {

    public PlusDMIndicatorTest(NumFactory numFactory) {
        super(numFactory);
    }

    @Test
    public void zeroDirectionalMovement() {
        var series = new MockBarSeriesBuilder().withNumFactory(numFactory).build();
        var yesterdayBar = series.barBuilder().openPrice(0).closePrice(0).highPrice(10).lowPrice(2).build();
        var todayBar = series.barBuilder().openPrice(0).closePrice(0).highPrice(6).lowPrice(6).build();
        series.addBar(yesterdayBar);
        series.addBar(todayBar);

        var dup = new PlusDMIndicator(series);
        assertNumEquals(0, dup.getValue(1));
    }

    @Test
    public void zeroDirectionalMovement2() {
        var series = new MockBarSeriesBuilder().withNumFactory(numFactory).build();
        var yesterdayBar = series.barBuilder().openPrice(0).closePrice(0).highPrice(6).lowPrice(12).build();
        var todayBar = series.barBuilder().openPrice(0).closePrice(0).highPrice(12).lowPrice(6).build();
        series.addBar(yesterdayBar);
        series.addBar(todayBar);

        var dup = new PlusDMIndicator(series);
        assertNumEquals(0, dup.getValue(1));
    }

    @Test
    public void zeroDirectionalMovement3() {
        var series = new MockBarSeriesBuilder().withNumFactory(numFactory).build();
        var yesterdayBar = series.barBuilder().openPrice(0).closePrice(0).highPrice(6).lowPrice(20).build();
        var todayBar = series.barBuilder().openPrice(0).closePrice(0).highPrice(12).lowPrice(4).build();
        series.addBar(yesterdayBar);
        series.addBar(todayBar);

        var dup = new PlusDMIndicator(series);
        assertNumEquals(0, dup.getValue(1));
    }

    @Test
    public void positiveDirectionalMovement() {
        var series = new MockBarSeriesBuilder().withNumFactory(numFactory).build();
        var yesterdayBar = series.barBuilder().openPrice(0).closePrice(0).highPrice(6).lowPrice(6).build();
        var todayBar = series.barBuilder().openPrice(0).closePrice(0).highPrice(12).lowPrice(4).build();
        series.addBar(yesterdayBar);
        series.addBar(todayBar);

        var dup = new PlusDMIndicator(series);
        assertNumEquals(6, dup.getValue(1));
    }
}

File: ta4j-core/src/test/kotlin/org/ta4j/core/indicators/adx/PlusDMIndicatorTest.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.indicators.adx

import org.junit.Test
import org.ta4j.core.Bar
import org.ta4j.core.Indicator
import org.ta4j.core.TestUtils
import org.ta4j.core.indicators.AbstractIndicatorTest
import org.ta4j.core.mocks.MockBar
import org.ta4j.core.mocks.MockBarSeries
import org.ta4j.core.num.Num
import java.util.function.Function

class PlusDMIndicatorTest(numFunction: Function<Number?, Num>) :
    AbstractIndicatorTest<Indicator<Num>?, Num>(numFunction) {
    @Test
    fun zeroDirectionalMovement() {
        val yesterdayBar = MockBar(0.0, 0.0, 10.0, 2.0, numFunction)
        val todayBar = MockBar(0.0, 0.0, 6.0, 6.0, numFunction)
        val bars: MutableList<Bar> = ArrayList()
        bars.add(yesterdayBar)
        bars.add(todayBar)
        val series = MockBarSeries(bars)
        val dup = PlusDMIndicator(series)
        TestUtils.assertNumEquals(0, dup.getValue(1))
    }

    @Test
    fun zeroDirectionalMovement2() {
        val yesterdayBar = MockBar(0.0, 0.0, 6.0, 12.0, numFunction)
        val todayBar = MockBar(0.0, 0.0, 12.0, 6.0, numFunction)
        val bars: MutableList<Bar> = ArrayList()
        bars.add(yesterdayBar)
        bars.add(todayBar)
        val series = MockBarSeries(bars)
        val dup = PlusDMIndicator(series)
        TestUtils.assertNumEquals(0, dup.getValue(1))
    }

    @Test
    fun zeroDirectionalMovement3() {
        val yesterdayBar = MockBar(0.0, 0.0, 6.0, 20.0, numFunction)
        val todayBar = MockBar(0.0, 0.0, 12.0, 4.0, numFunction)
        val bars: MutableList<Bar> = ArrayList()
        bars.add(yesterdayBar)
        bars.add(todayBar)
        val series = MockBarSeries(bars)
        val dup = PlusDMIndicator(series)
        TestUtils.assertNumEquals(0, dup.getValue(1))
    }

    @Test
    fun positiveDirectionalMovement() {
        val yesterdayBar = MockBar(0.0, 0.0, 6.0, 6.0, numFunction)
        val todayBar = MockBar(0.0, 0.0, 12.0, 4.0, numFunction)
        val bars: MutableList<Bar> = ArrayList()
        bars.add(yesterdayBar)
        bars.add(todayBar)
        val series = MockBarSeries(bars)
        val dup = PlusDMIndicator(series)
        TestUtils.assertNumEquals(6, dup.getValue(1))
    }
}